### PR TITLE
Implement `cupyx.scipy.linalg.solve_triangular`

### PR DESCRIPTION
**Pull request by YoshikawaMasashi**
I am implementing `cupyx.scipy.linalg.solve_triangular`.
https://docs.scipy.org/doc/scipy/reference/generated/scipy.linalg.solve_triangular.html
chainer/chainer#4899

### CODE REVIEW

**Comment by YoshikawaMasashi** [comment]
Implemented

File: cupyx/scipy/linalg/solve_triangular.py [hunk 1]
```diff
@@ -0,0 +1,103 @@
...
+            The matrix with dimension ``(M,)`` or ``(M, N)``.
+
+    .. seealso:: :func:`scipy.linalg.solve_triangular`
+    """
+
+    util._assert_cupy_array(a, b)
+
+    if len(a.shape) != 2 or a.shape[0] != a.shape[1]:
+        raise ValueError('expected square matrix')
+    if a.shape[0] != b.shape[0]:
```
Comment (okuta) [inline cupyx/scipy/linalg/solve_triangular.py]: Please use `if len(a) != len(b)`.

File: cupyx/scipy/linalg/solve_triangular.py [hunk 2]
```diff
@@ -0,0 +1,103 @@
...
+        raise ValueError('incompatible dimensions')
+
+    # Cast to float32 or float64
+    if a.dtype.char == 'f' or a.dtype.char == 'd':
```
Comment (okuta) [inline cupyx/scipy/linalg/solve_triangular.py]: How about `a.dtype.char in 'fd'` ?

File: cupyx/scipy/linalg/solve_triangular.py [hunk 3]
```diff
@@ -0,0 +1,103 @@
...
+        dtype = a.dtype
+    else:
+        dtype = numpy.find_common_type((a.dtype.char, 'f'), ())
+
+    a = a.astype(dtype)
```
Comment (okuta) [inline cupyx/scipy/linalg/solve_triangular.py]: Please use `a = cupy.array(a, dtype=dtype, order='F', copy=False)`.

File: cupyx/scipy/linalg/solve_triangular.py [hunk 4]
```diff
@@ -0,0 +1,103 @@
...
+    b = b.astype(dtype)
```
Comment (okuta) [inline cupyx/scipy/linalg/solve_triangular.py]: ditto.

File: cupyx/scipy/linalg/solve_triangular.py [hunk 5]
```diff
@@ -0,0 +1,103 @@
...
+
+    if check_finite:
+        if a.dtype.char in typecodes['AllFloat'] and \
```
Comment (okuta) [inline cupyx/scipy/linalg/solve_triangular.py]: Please use `a.dtype.kind =='f'`.

File: cupyx/scipy/linalg/solve_triangular.py [hunk 6]
```diff
@@ -0,0 +1,103 @@
...
+           not cupy.isfinite(a).all():
+            raise ValueError(
+                "array must not contain infs or NaNs")
+        if b.dtype.char in typecodes['AllFloat'] and \
```
Comment (okuta) [inline cupyx/scipy/linalg/solve_triangular.py]: ditto.

File: cupyx/scipy/linalg/solve_triangular.py [hunk 7]
```diff
@@ -0,0 +1,103 @@
...
+from numpy.core.numerictypes import typecodes
+
+import cupy
+from cupy.cuda import cublas
+from cupy.cuda import device
+from cupy.linalg import util
+
+
+def solve_triangular(a, b, trans=0, lower=False, unit_diagonal=False,
+                     overwrite_b=False, check_finite=True):
```
Comment (okuta) [inline cupyx/scipy/linalg/solve_triangular.py]: Do users expect finite checking by default?

File: cupyx/scipy/linalg/solve_triangular.py [hunk 8]
```diff
@@ -0,0 +1,103 @@
...
+        if a.dtype.char in typecodes['AllFloat'] and \
+           not cupy.isfinite(a).all():
+            raise ValueError(
+                "array must not contain infs or NaNs")
+        if b.dtype.char in typecodes['AllFloat'] and \
+           not cupy.isfinite(b).all():
+            raise ValueError(
+                "array must not contain infs or NaNs")
+
+    a = cupy.asfortranarray(a)
```
Comment (okuta) [inline cupyx/scipy/linalg/solve_triangular.py]: It can be removed. Please check another comment.

File: cupyx/scipy/linalg/solve_triangular.py [hunk 9]
```diff
@@ -0,0 +1,103 @@
...
+    b = cupy.asfortranarray(b)
```
Comment (okuta) [inline cupyx/scipy/linalg/solve_triangular.py]: It can be removed.

File: cupyx/scipy/linalg/solve_triangular.py [hunk 10]
```diff
@@ -0,0 +1,103 @@
...
+
+    dtype = a.dtype
```
Comment (okuta) [inline cupyx/scipy/linalg/solve_triangular.py]: Please remove.

File: cupyx/scipy/linalg/solve_triangular.py (hunk 7 above)
Comment (YoshikawaMasashi) [reply cupyx/scipy/linalg/solve_triangular.py]: OK. I rewrite. `check_finite=False`

**Comment by okuta** [comment]
jenkins, test this please.

**Comment by okuta** [comment]
LGTM!

**Pull request merged**